webap/seechange_webap.py

```diff
@@ -84,46 +102,114 @@ def exposures():
+                  'INNER JOIN provenance_tags ept ON ept.provenance_id=e.provenance_id AND ept.tag=%(provtag)s '
+                  'GROUP BY e.id, i.id, s.id, sl.id, sl.num_sources '
+                 )
+            subdict['provtag'] = data['provenancetag']
         if ( t0 is not None ) or ( t1 is not None ):
-            q += "  WHERE "
+            q += 'WHERE '
             if t0 is not None:
                 q += 'e.mjd >= %(t0)s'
                 subdict['t0'] = t0
             if t1 is not None:
                 if t0 is not None: q += ' AND '
                 q += 'e.mjd <= %(t1)s'
                 subdict['t1'] = t1
-        q += ( '   GROUP BY e.id ' # ,e.filepath,e.mjd,e.target,e.filter,e.filter_array,e.exp_time '
-               '   ORDER BY e.mjd, e.filter, e.filter_array ' )
-
-        q += ( ') m '
-               'LEFT JOIN reports r ON m.id=r.exposure_id '
-               'GROUP BY m.id, m.filepath, m.mjd, m.target, m.filter, m.filter_array, m.exp_time, '
-               '         m.n_images, m.n_cutouts, m.n_measurements ' )
 
         cursor.execute( q, subdict )
+
+        # Now run a second query to count and sum those things
+        # These numbers will be wrong (double-counts) if not filtering on a provenance tag, or if the
+        #   provenance tag includes multiple provenances for a given step!
+        q = ( 'SELECT t.id, t.filepath, t.mjd, t.target, t.filter, t.filter_array, t.exp_time, '
+              '  COUNT(t.subid) AS num_subs, SUM(t.num_sources) AS num_sources, '
+              '  SUM(t.num_measurements) AS num_measurements '
+              'INTO TEMP TABLE temp_imgs_2 '
+              'FROM temp_imgs t '
+              'GROUP BY t.id, t.filepath, t.mjd, t.target, t.filter, t.filter_array, t.exp_time '
+             )
+
+        cursor.execute( q )
+
+        # Run a third query count reports
+        subdict = {}
+        q = ( 'SELECT t.id, t.filepath, t.mjd, t.target, t.filter, t.filter_array, t.exp_time, '
+              '  t.num_subs, t.num_sources, t.num_measurements, '
+              '  SUM( CASE WHEN r.success THEN 1 ELSE 0 END ) as n_successim, '
+              '  SUM( CASE WHEN r.error_message IS NOT NULL THEN 1 ELSE 0 END ) AS n_errors '
+              'FROM temp_imgs_2 t '
+             )
+        if data['provenancetag'] is None:
+            q += 'LEFT JOIN reports r ON r.exposure_id=t.id '
+        else:
+            q += ( 'LEFT JOIN ( '
+                   '  SELECT re.exposure_id, re.success, re.error_message '
+                   '  FROM reports re '
+                   '  INNER JOIN provenance_tags rept ON rept.provenance_id=re.provenance_id AND rept.tag=%(provtag)s '
+                   ') r ON r.exposure_id=t.id '
+                  )
+            subdict['provtag'] = data['provenancetag']
+        # I wonder if making a primary key on the temp table would be more efficient than
+        #    all these columns in GROUP BY?  Investigate this.
```

Comment:
I feel like I could take an entire day just studying this monstrous 3-part query! sqlalchemy has shielded me from pure SQL, mostly...

Reply:
Yeah, these things get kind of nasty.

Adding the provenance tags made these queries much more complicated.